src/frontend/packages/cf-autoscaler/src/core/autoscaler-helpers/autoscaler-transform-metric.ts

```diff
@@ -0,0 +1,301 @@
+}
+
+function buildMetricColorData(metricData: AppAutoscalerMetricDataPoint[], trigger: AppScalingTrigger): AppAutoscalerMetricDataPoint[] {
+  const colorTarget: AppAutoscalerMetricDataPoint[] = [];
+  metricData.map((item) => {
+    colorTarget.push({
+      name: item.name,
+      value: getColor(trigger, item.value),
+    });
+  });
+  if (trigger.upper && trigger.upper.length > 0) {
+    buildSingleColor(colorTarget, trigger.upper);
+  }
+  if (trigger.lower && trigger.lower.length > 0) {
+    buildSingleColor(colorTarget, trigger.lower);
+  }
+  return colorTarget;
+}
+
+function buildSingleColor(lineChartSeries: AppAutoscalerMetricDataPoint[], ul: AppScalingRule[]) {
+  ul.forEach((item) => {
+    const lineData = {
+      name: buildTriggerName(item),
+      value: item.color
+    };
+    lineChartSeries.push(lineData);
+  });
+}
+
+function buildMarkLineData(metricData, trigger: AppScalingTrigger): AppAutoscalerMetricDataLine[] {
+  const lineChartSeries: AppAutoscalerMetricDataLine[] = [];
+  if (trigger.upper && trigger.upper.length > 0) {
+    buildSingleMarkLine(lineChartSeries, metricData, trigger.upper);
+  }
+  if (trigger.lower && trigger.lower.length > 0) {
+    buildSingleMarkLine(lineChartSeries, metricData, trigger.lower);
+  }
+  return lineChartSeries;
+}
+
+function buildTriggerName(item: AppScalingRule): string {
+  const type = getScaleType(item.operator);
+  return `${type} threshold: ${item.operator} ${item.threshold}`;
+}
+
+function buildSingleMarkLine(lineChartSeries: AppAutoscalerMetricDataLine[], metricData: AppAutoscalerMetricData[], ul: AppScalingRule[]) {
+  ul.forEach(item => {
+    const lineData = {
+      name: buildTriggerName(item),
+      series: []
+    };
+    metricData.map((data) => {
+      lineData.series.push({
+        name: data.name,
+        value: item.threshold,
+      });
+    });
+    lineChartSeries.push(lineData);
+  });
+}
+
+function executeCompare(val1: number, operator: string, val2: number): boolean {
+  switch (operator) {
+    case '>':
+      return val1 > val2;
+    case '>=':
+      return val1 >= val2;
+    case '<':
+      return val1 < val2;
+    default:
+      return val1 <= val2;
+  }
+}
+
+function getColor(trigger: AppScalingTrigger, value: any): string {
+  let color = AutoscalerConstants.normalColor;
+  if (Number.isNaN(value)) {
+    return color;
+  }
+  for (let i = 0; trigger.upper && trigger.upper.length > 0 && i < trigger.upper.length; i++) {
+    if (executeCompare(value, trigger.upper[i].operator, trigger.upper[i].threshold)) {
+      color = trigger.upper[i].color;
+      break;
+    }
+  }
+  for (let i = 0; trigger.lower && trigger.lower.length > 0 && i < trigger.lower.length; i++) {
+    const index = trigger.lower.length - 1 - i;
+    if (executeCompare(value, trigger.lower[index].operator, trigger.lower[index].threshold)) {
+      color = trigger.lower[index].color;
+      break;
+    }
+  }
```

Comment:
The for loop may not output a color value in some of the situation, so if we extracted the for loop to a new function `getColorFromRule()`, we will have to introduce a new variable `const newColor = getColorFromRule(...)` and check the whether the `newColor` is an empty string... it's more complicated...
So I used another way to avoid the value-change of `color` and define it as a const.

Reply:
@KlapTrap